File: artifacts/api-server/src/routes/whatsapp.ts
import { Router, type Request, type Response } from "express";
import { SectorsService } from "../services/sectors.service.js";
import { StocksService } from "../services/stocks.service.js";
import { PatternsService } from "../services/patterns.service.js";
import { ScannersService } from "../services/scanners.service.js";
import { WhatsappService } from "../services/whatsapp.service.js";
import { NseService } from "../services/nse.service.js";
import { YahooService } from "../services/yahoo.service.js";

const router = Router();

let whatsappService: WhatsappService | null = null;
function getService() {
  if (!whatsappService) {
    const nse     = new NseService();
    const yahoo   = new YahooService();
    const sectors  = new SectorsService(nse, yahoo);
    const stocks   = new StocksService(nse, yahoo);
    const patterns = new PatternsService(yahoo, nse);
    const scanners = new ScannersService(yahoo, nse);
    whatsappService = new WhatsappService(sectors, stocks, patterns, scanners);
  }
  return whatsappService;
}

// ─── Simple in-memory rate limiter (per IP, per minute) ──────────────────────

const msgCounts = new Map<string, { count: number; resetAt: number }>();
const MSG_LIMIT = 10;

function isRateLimited(ip: string): boolean {
  const now = Date.now();
  const entry = msgCounts.get(ip);
  if (!entry || now > entry.resetAt) {
    msgCounts.set(ip, { count: 1, resetAt: now + 60_000 });
    return false;
  }
  entry.count += 1;
  if (entry.count > MSG_LIMIT) return true;
  return false;
}

// Periodically clean up stale rate-limit entries to avoid unbounded growth
setInterval(() => {
  const now = Date.now();
  for (const [k, v] of msgCounts) if (now > v.resetAt) msgCounts.delete(k);
}, 60_000).unref();

// ─── Routes ───────────────────────────────────────────────────────────────────

router.get("/status", (_req: Request, res: Response) => {
  res.json(getService().getBotStatus());
});

router.post("/message", async (req: Request, res: Response) => {
  const ip = (req.headers["x-forwarded-for"] as string | undefined)?.split(",")[0].trim()
    ?? req.socket.remoteAddress
    ?? "unknown";

  if (isRateLimited(ip)) {
    res.status(429).json({ error: "Too many requests. Please wait a minute." });
    return;
  }

  const { from, message } = req.body ?? {};
  if (!message || typeof message !== "string" || !message.trim()) {
    res.status(400).json({ error: "message field is required" });
    return;
  }

  try {
    const response = await getService().processMessage({ from: from ?? "web-user", message });
    res.json(response);
  } catch (err) {
    req.log.error({ err }, "Bot message processing failed");
    res.status(500).json({ error: "Failed to process message" });
  }
});

/** Twilio webhook — paste as "When a message comes in" URL in Twilio console */
router.post("/twilio", async (req: Request, res: Response) => {
  const body = req.body as Record<string, string>;
  const messageText = body.Body || body.body || "";
  const from        = body.From || body.from || "unknown";

  try {
    const result = await getService().processMessage({ from, message: messageText });
    const replyText = result.response || "No response";

    const escaped = replyText
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;")
      .replace(/"/g, "&quot;");

    res.set("Content-Type", "text/xml");
    res.send(`<?xml version="1.0" encoding="UTF-8"?>
<Response>
  <Message>${escaped}</Message>
</Response>`);
  } catch (err) {
    req.log.error({ err }, "Twilio webhook failed");
    res.set("Content-Type", "text/xml");
    res.send(`<?xml version="1.0" encoding="UTF-8"?>
<Response>
  <Message>Sorry, something went wrong. Please try again.</Message>
</Response>`);
  }
});

router.get("/messages", (_req: Request, res: Response) => {
  res.json(getService().getMessageLog());
});

router.post("/qr", (_req: Request, res: Response) => {
  res.json(getService().simulateQrCode());
});

router.put("/status", (req: Request, res: Response) => {
  const { enabled } = req.body ?? {};
  if (typeof enabled !== "boolean") {
    res.status(400).json({ error: "enabled must be a boolean" });
    return;
  }
  res.json(getService().updateBotStatus(enabled));
});

export default router;

File: artifacts/python-backend/app/routes/whatsapp.py
import hmac
import hashlib
import base64
import logging
import os

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse, Response
from typing import Any

from ..services.whatsapp_service import WhatsappService
from ..services.sectors_service import SectorsService
from ..services.stocks_service import StocksService
from ..services.patterns_service import PatternsService
from ..services.scanners_service import ScannersService
from ..services.nse_service import NseService
from ..services.yahoo_service import YahooService
from ..services.price_service import PriceService
from ..services.nlp_service import NlpService
from ..services.bot_dispatcher import BotDispatcher
from ..services import news_service as _news_module

logger = logging.getLogger(__name__)


def _verify_twilio_signature(request: Request, form: dict[str, str]) -> bool:
    """Verify the X-Twilio-Signature header against the Twilio auth token.

    Returns True if the signature matches, False otherwise. If TWILIO_AUTH_TOKEN
    is not configured, returns False — fail closed rather than accept any
    payload as if it came from Twilio.

    Algorithm (per Twilio docs): HMAC-SHA1 of (full request URL +
    sorted-key concatenation of form params), base64 encoded.
    """
    auth_token = os.environ.get("TWILIO_AUTH_TOKEN", "")
    if not auth_token:
        return False
    received = request.headers.get("X-Twilio-Signature", "")
    if not received:
        return False
    # Reconstruct the URL Twilio used. If you sit behind a proxy that
    # rewrites the host/scheme, set TWILIO_WEBHOOK_URL to the public URL.
    url = os.environ.get("TWILIO_WEBHOOK_URL", "").strip()
    if not url:
        url = str(request.url)
    # Concatenate form params sorted by key (Twilio's signing recipe).
    data = url + "".join(f"{k}{form[k]}" for k in sorted(form.keys()))
    digest = hmac.new(
        auth_token.encode("utf-8"),
        data.encode("utf-8"),
        hashlib.sha1,
    ).digest()
    expected = base64.b64encode(digest).decode("utf-8")
    return hmac.compare_digest(expected, received)

router = APIRouter(prefix="/whatsapp", tags=["whatsapp"])

_nse      = NseService()
_yahoo    = YahooService()
_price    = PriceService(_nse, _yahoo)
_sectors  = SectorsService(_nse, _yahoo)
_stocks   = StocksService(_nse, _yahoo)
_patterns = PatternsService(_yahoo, _nse)
_scanners = ScannersService(_price)
_nlp      = NlpService()

try:
    from ..services.hydra_service import HydraEngine
    _hydra = HydraEngine()
except Exception:  # pragma: no cover
    _hydra = None

_dispatcher = BotDispatcher(
    sectors=_sectors, stocks=_stocks, patterns=_patterns, scanners=_scanners,
    nlp=_nlp, hydra=_hydra, news=_news_module,
)
_service = WhatsappService(_sectors, _stocks, _patterns, _scanners, _nlp,
                           dispatcher=_dispatcher)


def get_service() -> WhatsappService:
    return _service


def get_dispatcher() -> BotDispatcher:
    return _dispatcher


@router.get("/status")
async def get_status():
    return _service.get_bot_status()


@router.put("/status")
async def update_status(body: dict[str, Any]):
    enabled = body.get("enabled")
    if enabled is None:
        return JSONResponse(status_code=400, content={"error": "'enabled' field required"})
    return _service.update_bot_status(bool(enabled))


@router.post("/status")
async def set_status(body: dict[str, Any]):
    enabled = body.get("enabled")
    if enabled is None:
        return JSONResponse(status_code=400, content={"error": "'enabled' field required"})
    return _service.update_bot_status(bool(enabled))


@router.post("/message")
async def process_message(body: dict[str, Any]):
    try:
        return await _service.process_message(body)
    except ValueError as e:
        return JSONResponse(status_code=400, content={"error": str(e)})


@router.post("/twilio")
async def twilio_webhook(request: Request):
    form = await request.form()
    # Twilio's form is a Starlette FormData; flatten str values for signing.
    form_dict = {k: str(v) for k, v in form.items() if isinstance(v, (str, bytes))}
    if not _verify_twilio_signature(request, form_dict):
        logger.warning("Twilio webhook: invalid X-Twilio-Signature, rejecting.")
        return JSONResponse(status_code=403, content={"error": "Forbidden"})
    from_number = form.get("From") or form.get("from") or "whatsapp:+unknown"
    text = form.get("Body") or form.get("body") or ""
    try:
        result = await _service.process_message(
            {"from": str(from_number), "text": str(text)}
        )
        reply = result.get("response") or "Sorry, I could not process your request."
    except Exception:
        logger.exception("Twilio webhook: processing failed")
        reply = "Sorry, I could not process your request."
    reply_safe = (
        str(reply)
        .replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
        .replace('"', "&quot;")
    )
    twiml = (
        '<?xml version="1.0" encoding="UTF-8"?>'
        "<Response>"
        f"<Message>{reply_safe}</Message>"
        "</Response>"
    )
    return Response(content=twiml, media_type="application/xml")


@router.get("/messages")
async def get_messages():
    return _service.get_message_log()


@router.post("/qr")
async def generate_qr():
    return _service.simulate_qr_code()
